File: 004/Ruby/problem004_test.rb
require "test/unit"

require "problem004"

class TestProblem004 < Test::Unit::TestCase
  def test_palindrome_huh
    assert 9.palindrome?
    assert 33.palindrome?
    assert 131.palindrome?
    assert 451121154.palindrome?
    assert 249942.palindrome?
  end
  
  def test_not_palindrome_huh
    assert ! 19.palindrome?
    assert ! 633.palindrome?
    assert ! 1231.palindrome?
    assert ! 4561121154.palindrome?
    assert ! 22499432.palindrome?
  end
  
  def test_find_palindrome_product
    assert_equal 9009, 10.find_palindrome_product
  end
  
end

File: 004/Haskell/Problem004Test.hs
module Main where

import Test.HUnit
import System.Exit
import Problem004

isPalindromeTests = TestList [
  TestCase (assertEqual "simple palindrome" True (isPalindrome 9)),
  TestCase (assertEqual "simple palindrome" True (isPalindrome 33)),
  TestCase (assertEqual "simple palindrome" True (isPalindrome 101)),
  TestCase (assertEqual "simple palindrome" True (isPalindrome 127848721)),
  TestCase (assertEqual "simple palindrome" True (isPalindrome 23233232)),

  TestCase (assertEqual "simple palindrome" False (isPalindrome 93)),
  TestCase (assertEqual "simple palindrome" False (isPalindrome 331)),
  TestCase (assertEqual "simple palindrome" False (isPalindrome 1019)),
  TestCase (assertEqual "simple palindrome" False (isPalindrome 12748721)),
  TestCase (assertEqual "simple palindrome" False (isPalindrome 23232323))
  ]

findPalindromeProductTests = TestList [
  TestCase (assertEqual "two-digit-factors palindrome" 9009 (findPalindromeProduct 10))
  ]

checkTests counts = case counts of
	Counts _ _ errors failures ->
		if (errors + failures > 0)
		then exitFailure
		else (exitWith ExitSuccess)

main = runTestTT
        (TestList [
          TestLabel "testing is palindrome" isPalindromeTests,
          TestLabel "testing palindrome search" findPalindromeProductTests
          ])
       >>= checkTests
